### PR TITLE
Don't count kill assist on build time or player immunity

### PR DESCRIPTION
**Pull request by savilli**
Fix https://github.com/MT-CTF/capturetheflag/issues/793
Introduce ctf.register_can_attack callbacks that can stop any ctf related attack logic.

### CODE REVIEW

File: mods/ctf/ctf/teams.lua [hunk 1]
```diff
@@ -446,12 +446,20 @@ function ctf.register_on_killedplayer(func)
...
-ctf.registered_on_punchplayer = {}
-function ctf.register_on_punchplayer(func)
+ctf.registered_can_attack = {}
+function ctf.register_can_attack(func)
 	if ctf._mt_loaded then
 		error("You can't register callbacks at game time!")
 	end
-	table.insert(ctf.registered_on_punchplayer, func)
+	table.insert(ctf.registered_can_attack, func)
+end
```
Comment (LoneWolfHT) [inline mods/ctf/ctf/teams.lua:455]: Just create a ctf.can_attack() function. Mods will still be able to add things to it if needed. Will result in clearer code too

**Review by LoneWolfHT — CHANGES_REQUESTED** [review]

File: mods/ctf/ctf/teams.lua (hunk 1 above)
Comment (savilli) [reply mods/ctf/ctf/teams.lua:455]: >  Mods will still be able to add things to it if needed
How so? What mod should do if it want to reject the damage?

Comment (LoneWolfHT) [reply mods/ctf/ctf/teams.lua:455]:
```lua
local oldfunc = ctf.can_attack
ctf.can_attack = function()
    -- New code here, can call the old function with oldfunc()
end
```

Comment (savilli) [reply mods/ctf/ctf/teams.lua:455]: We already have a common way to register callback using register_ functions. It's used in minetest api as well as in ctf (ctf.register_on_killedplayer, ctf.register_on_punchplayer). Why you want to do it another way? In my view two different ways to register callback will only add confusion and decrease code readability.

Comment (LoneWolfHT) [reply mods/ctf/ctf/teams.lua:455]: The register_on() callbacks are used to let mods change behavior of something else. In this case you want a mod to change it's behavior based on what something else says. In a situation like that you use the method I just mentioned.
See minetest.is_protected() for an example. CTF also has a few functions doing the same. Although I forget where exactly they are.

Comment (savilli) [reply mods/ctf/ctf/teams.lua:455]: It's absolutely not clear how to distinguish two cases. But minetest.is_protected is a valid case. If you insist on that way I'll do like is_protected.

**Pull request merged**